snap/integrity/integrity.go

```diff
@@ -48,6 +55,100 @@ func align(size uint64) uint64 {
 	return (size + blockSize - 1) / blockSize * blockSize
 }
 
+type IntegrityData struct {
+	Header         *IntegrityDataHeader
+	Offset         uint64
+	SourceFilePath string
+
+	Bytes *[]byte
+}
+
+func FindIntegrityData(snapName string) (*IntegrityData, error) {
+	integrityData := IntegrityData{}
+
+	if !squashfs.FileHasSquashfsHeader(snapName) {
+		return nil, fmt.Errorf("input file does not contain a SquashFS filesystem")
+	}
+
+	snapFile, err := os.OpenFile(snapName, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer snapFile.Close()
+
+	// Seek to bytes_used field of SquashFS superblock
+	_, err = snapFile.Seek(squashfsSuperblockBytesUsedOffset, io.SeekStart)
+
+	var squashFSSize uint64
+	if err := binary.Read(snapFile, binary.LittleEndian, &squashFSSize); err != nil {
+		return nil, err
+	}
+
+	logger.Debugf("SquashFS bytes used: %d", squashFSSize)
+
+	snapFileInfo, err := snapFile.Stat()
+	if err != nil {
+		return nil, err
+	}
+
+	// Align squashFSSize to blockSize
+	offset := align(squashFSSize)
+
+	// If no extra data attached at the end of the squashfs return empty integrityData
+	if offset == uint64(snapFileInfo.Size()) {
+		return nil, nil
+	}
+
+	_, err = snapFile.Seek(int64(offset), io.SeekStart)
+
+	integrityDataBytes := make([]byte, uint64(snapFileInfo.Size())-offset)
+	n, err := snapFile.Read(integrityDataBytes)
```

Comment:
as I mentioned before I don't think we want to read all this data into memory

Reply:
This is only reading the 4096 byte header from the end of the snap, not the entire verity data. Is it worth using a buffered read for this? In reality the read here could be much less than 4096 as the integrity header contains just a string ("integrity"), a root hash and the total size.

Reply:
it seems I looked at an old version of the code?

Reply:
Possibly, I can find the calculation `snapFileInfo.Size() - offset` in the first commit in this branch. In the commit after that I changed that to use the new const `HeaderSize`